app/Community/Actions/SendDailyDigestAction.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Community\Actions;
+
+use App\Community\Enums\SubscriptionSubjectType;
+use App\Mail\DailyDigestMail;
+use App\Models\ForumTopic;
+use App\Models\ForumTopicComment;
+use App\Models\User;
+use App\Models\UserDelayedSubscription;
+use Illuminate\Support\Facades\Mail;
+
+class SendDailyDigestAction
+{
+    public function execute(User $user): void
+    {
+        // load up the pending subscriptions, then delete them.
+        $delayedSubscriptions = UserDelayedSubscription::query()
+            ->where('user_id', $user->id)
+            ->orderBy('id')
+            ->get();
+        $last = $delayedSubscriptions->last();
+        if (!$last) {
+            return;
+        }
+        UserDelayedSubscription::query()
+            ->where('user_id', $user->id)
+            ->where('id', '<=', $last->id)
+            ->delete();
+
+        // if the user doesn't have an email address, bail.
+        // do this after deleting the pending subscriptions.
+        if (!$user->EmailAddress) {
+            return;
+        }
+
+        // build a map of ids for each type
+        $ids = [];
+        foreach ($delayedSubscriptions as $delayedSubscription) {
+            if (!isset($ids[$delayedSubscription->subject_type->value])) {
+                $ids[$delayedSubscription->subject_type->value] = [];
+            }
+            $ids[$delayedSubscription->subject_type->value][] = $delayedSubscription->subject_id;
+        }
+        // and preload the titles for the associated records
+        $titles = [];
+        foreach ($ids as $type => $typeIds) {
+            $titles[$type] = match ($type) {
+                SubscriptionSubjectType::ForumTopic->value => ForumTopic::whereIn('id', $typeIds)->pluck('title', 'id'),
+                default => [],
+            };
+        }
+
+        // build the data to pass to the mail script
+        $notificationItems = [];
+        foreach ($delayedSubscriptions as $delayedSubscription) {
+            // if all the new posts have been deleted or aren't visible, ignore it
+            $handler = $this->getHandler($delayedSubscription->subject_type);
+            $count = $handler->getUpdatesSince($delayedSubscription);
+            if ($count > 0) {
+                $notificationItems[] = [
+                    'type' => $delayedSubscription->subject_type->value,
+                    'title' => $titles[$delayedSubscription->subject_type->value][$delayedSubscription->subject_id] ?? '(untitled)',
+                    'link' => $handler->getLink($delayedSubscription->subject_id, $delayedSubscription->first_update_id),
+                    'count' => $count,
+                ];
+            }
+        }
+
+        // send the mail
+        Mail::to($user->EmailAddress)->queue(
+            new DailyDigestMail($user, $notificationItems)
+        );
+    }
```

Comment:
I could be misinterpreting, but I _think_ this logic means it's theoretically possible for `$notificationItems` to be empty and, thus, our email body to virtually be empty.

Even if that's implausible, we could save ourselves a lot of trouble in the long run with overzealous email providers by:
```php
if (empty($notificationItems)) {
    return;
}

// send the mail
Mail::to(...);
```

Reply:
Possible, but unlikely. It would mean every delayed subscription the user was queued up for got deleted before the email was sent. I've added the return anyway.